docs/madr/decisions/033-meshhttproute-meshgateway.md

```diff
@@ -0,0 +1,338 @@
+host matching separate is very high. It also differs from Gateway API, which is
+a disadvantage.
+
+### `to.targetRef`
+
+When using `MeshHTTPRoute`/`MeshTCPRoute` for service to service communication, it's very
+useful to limit the effects based on request source and destination. We do
+this by setting `spec.targetRef`, the source, and `spec.to[].targetRef`, the destination.
+
+With `MeshGateway`, once a request hits the gateway, the routing resources
+_define_ the destination. We have to put the rules somewhere, so this MADR
+proposes `spec.to`. There's no real semantic value to a distinct
+`spec.to[].targetRef`, or for that matter, a `spec.from[].targetRef`.
+
+This MADR proposes requiring `spec.to[].targetRef.kind: Mesh`.
+
+Allowing `to[].targetRef` to be empty would be a break from all other policies.
+Eventually we can support it generally via a default value in the open API schema,
+see also [#6070](https://github.com/kumahq/kuma/issues/6070).
+
+### Route matching
+
+This MADR doesn't address supporting policy matching of routes
+that attach to `MeshGateways` because we can't policy match on `MeshGatewayRoute` at
+the moment anyway. The feasability/details should be addressed in a separate MADR.
+
+### `MeshHTTPRoute` example
+
+```
+apiVersion: kuma.io/v1alpha1
+kind: MeshHTTPRoute
+metadata:
+  name: edge-routes
+spec:
+  targetRef:
+    kind: MeshGateway
+    name: edge-gateway
+  to:
+  - targetRef:
+      kind: Mesh
+    hostnames:
+    - example.com
+    rules:
+    - matches:
+      - path:
+          type: PathPrefix
+          value: /
+      default:
+        backendRefs:
+        - kind: MeshService
+          name: backend_demo_svc_8080
+```
+
+### `MeshTCPRoute` example
+
+```
+apiVersion: kuma.io/v1alpha1
+kind: MeshTCPRoute
+metadata:
+  name: edge-routes
+spec:
+  targetRef:
+    kind: MeshGateway
+    name: edge-gateway
+  to:
+  - targetRef:
+      kind: Mesh
+    rules:
+      - default:
+          backendRefs:
+            - kind: MeshServiceSubset
+              name: backend_kuma-demo_svc_3001
+              tags:
+                version: "1.0"
+              weight: 90
+            - kind: MeshServiceSubset
+              name: backend_kuma-demo_svc_3001
+              tags:
+                version: "2.0"
+              weight: 10
+```
+
+### Merging example
+
+```
+---
+apiVersion: kuma.io/v1alpha1
+kind: MeshHTTPRoute
+metadata:
+  name: specific-hostname
+spec:
+  targetRef:
+    kind: MeshGateway
+    name: edge-gateway
+  to:
+  - targetRef:
+      kind: Mesh
+    hostnames:
+    - test.example.com
+    rules:
+    - matches:
+      - path:
+          type: PathPrefix
+          value: /
+      default:
+        backendRefs:
+        - kind: MeshServiceSubset
+          name: backend_demo_svc_8080
+          tags:
+            version: v1
+---
+apiVersion: kuma.io/v1alpha1
+kind: MeshHTTPRoute
+metadata:
+  name: wildcard-hostname
```

Comment:
do we care about the lexicographic order of the names in this case?

Reply:
Of the resource names? Don't we use them now to sort the rules?

Reply:
Yes, but it seems like for MeshHTTPRoute we build the order that makes the most sense, no? i.e. generic matches like `PathPrefix: /` on top and more specific `Exect: /bar` at the bottom?

Reply:
Yes this is already part of the MeshHTTPRoute spec independent of this PR.